### PR TITLE
Optimize JoinCondition matching

### PR DESCRIPTION
**Pull request by suneet-s**
### Description

The LookupJoinMatcher needs to check if a condition is always true or false
multiple times. This can be pre-computed to speed up the match checking

This change reduces the time it takes to perform a for joining on a long key
from ~ 36 ms/op to 23 ms/ op

Screen Shot 2020-01-16 at 3 34 16 PM

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/segment/join/JoinConditionAnalysis.java [hunk 1]
```diff
@@ -133,26 +142,23 @@ public String getOriginalExpression()
...
 
   /**
    * Return whether this condition is a constant that is always true.
    */
   public boolean isAlwaysTrue()
   {
-    return equiConditions.isEmpty() &&
-           nonEquiConditions.stream()
-                            .allMatch(expr -> expr.isLiteral() && expr.eval(ExprUtils.nilBindings()).asBoolean());
+    return equiConditions.isEmpty() && allTrueLiteralNonEquiConditions;
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/segment/join/JoinConditionAnalysis.java]: It seems like `allTrueLiteralNonEquiConditions` is only used here; how about caching `isAlwaysTrue` directly?

File: processing/src/main/java/org/apache/druid/segment/join/JoinConditionAnalysis.java [hunk 2]
```diff
@@ -133,26 +142,23 @@ public String getOriginalExpression()
    */
   public boolean isAlwaysFalse()
   {
-    return nonEquiConditions.stream()
-                            .anyMatch(expr -> expr.isLiteral() && !expr.eval(ExprUtils.nilBindings()).asBoolean());
+    return anyFalseLiteralNonEquiConditions;
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/segment/join/JoinConditionAnalysis.java]: Why not call this `isAlwaysFalse`? (It looks like it isn't used anywhere else, and it seems to me to be easier to understand the meaning of the field if it's named after what we want it to mean.)

Comment (suneet-s) [reply processing/src/main/java/org/apache/druid/segment/join/JoinConditionAnalysis.java]: yeah. I wasn't sure if isAlwaysFalse could be more complex going forward. This is clearer - it took me a while to wrap my head around what isAlwaysFalse was trying to check. Added comments, and used `isAlways...`

File: processing/src/main/java/org/apache/druid/segment/join/JoinConditionAnalysis.java (hunk 1 above)
Comment (suneet-s) [reply processing/src/main/java/org/apache/druid/segment/join/JoinConditionAnalysis.java]: Done

**Comment by suneet-s** [comment]
> Cool, thank you for the PR. It seems like the benchmark class is not in master yet. Would you please link the benchmark class here?

@jihoonson  here's the benchmark I was using. I will push it up to master once I clean it up a little more and add more tests
https://github.com/gianm/druid/blob/joins/benchmarks/src/main/java/org/apache/druid/benchmark/JoinAndLookupBenchmark.java

**Pull request merged**